File: ml-benchmark/autolifter/dac/3rd-min.f
Config ExtraGrammar = "AutoLifter";
Config ComposeNum = 2;
Config TermNum = 6;

Inductive List = cons {Int, List} | nil Unit;

length = fix (\f: List->Int. \x: List.
  match x with
    cons {h, t} -> + (f t) 1
  | nil _ -> 0
  end
);

concat = fix (\f: List->List->List. \x: List. \y: List.
  match x with
    cons {h, t} -> cons {h, f t y}
  | nil _ -> y
  end
);

split = \xs: List. fix (\f: List->Int->{List, List}. \x: List. \n: Int.
  if (< n 1) then {nil unit, x}
  else match x with
    cons {h, t} -> let res = f t (- n 1) in {cons {h, res.1}, res.2}
  | _ -> {nil unit, nil unit}
  end
) xs (/ (length xs) 2);

dac = \v: List -> Int. \xs: List.
  let run = fix (\f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {_, nil _} -> xs
    | _ -> let sp = split xs in
      concat (f sp.1) (f sp.2)
    end
  ) in v (run xs);

inf = 100;

/*User provided programs*/

min = \a: Int. \b: Int. if < a b then a else b;
max = \a: Int. \b: Int. if > a b then a else b;

trdmin = \xs: List. (fix (
  \f: List -> {Int, Int, Int}. \xs: List.
  match xs with
    nil _ -> {inf, inf, inf}
  | cons {h, t} ->
    let res = f t in
      {min res.1 (max res.2 h),
       min res.2 (max res.3 h),
       min res.3 h}
  end
) xs).1;

main = dac trdmin;

File: ml-benchmark/autolifter/dac/3rd-min.ml
type l = Cons of int * l | Nil of unit

let rec length x =
  match x with
  | Cons (h, t) -> length t + 1
  | Nil _ -> 0

let rec concat x y =
  match x with
  | Cons (h, t) -> Cons (h, concat t y)
  | Nil _ -> y

let split xs =
  let rec f x n =
    if n < 1 then (Nil (), x)
    else
      match x with
      | Cons (h, t) ->
        let res = f t (n - 1) in
        (match res with
         | (a, b) -> (Cons (h, a), b))
      | _ -> (Nil (), Nil ())
  in
  f xs (length xs / 2)

let dac v =
  fun xs ->
    let run =
      let rec f xs =
        match xs with
        | Nil _ -> xs
        | Cons (h, t) ->
          (match t with
           | Nil _ -> xs
           | _ ->
             let sp = split xs in
             (match sp with
              | (a, b) -> concat (f a) (f b)))
      in
      f
    in
    v (run xs)

let inf = 100

let min a b = if a < b then a else b
let max a b = if a > b then a else b

let trdmin xs =
  let rec f xs =
    match xs with
    | Nil _ -> (inf, inf, inf)
    | Cons (h, t) ->
      let res = f t in
      (match res with
       | (a, b, c) ->
         (min a (max b h),
          min b (max c h),
          min c h))
  in
  match f xs with
  | (a, b, c) -> a

let program = dac trdmin